pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +255,266 @@ func (p *Pod) Skip() bool {
 	return false
 }
 
+func (p *Pod) groupName() string {
+	return p.Object().GetLabels()[GroupNameLabel]
+}
+
+func (p *Pod) groupTotalCount() (int, error) {
+	if p.groupName() == "" {
+		return 0, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+
+	gtcAnnotation, ok := p.Object().GetAnnotations()[GroupTotalCountAnnotation]
+	if !ok {
+		return 0, fmt.Errorf("failed to extract '%s' annotation",
+			GroupTotalCountAnnotation)
+	}
+
+	gtc, err := strconv.Atoi(gtcAnnotation)
+	if err != nil {
+		return 0, err
+	}
+
+	if gtc < 1 {
+		return 0, fmt.Errorf("incorrect annotation value '%s=%s': group total count should be greater than zero",
+			GroupTotalCountAnnotation, gtcAnnotation)
+	}
+
+	return gtc, nil
+}
+
+func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet, error) {
+	groupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		resultPodSets []kueue.PodSet
+	)
+
+	for _, podInGroup := range podsInGroup.Items {
+		tc, err := strconv.Atoi(podInGroup.GetAnnotations()[GroupTotalCountAnnotation])
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
+		}
+		if tc != groupTotalCount {
+			// ToDo: Add a warning event
+			return nil, fmt.Errorf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc)
+		}
+
+		roleHash, ok := podInGroup.Annotations[RoleHashAnnotation]
+		if !ok {
+			roleHash, err = getRoleHash(fromObject(&podInGroup))
+		}
+
+		roleHash = roleHash[:8]
+
+		podRoleFound := false
+		for psi := range resultPodSets {
+			if resultPodSets[psi].Name == roleHash {
+				podRoleFound = true
+				resultPodSets[psi].Count++
+			}
+		}
+
+		if !podRoleFound {
+			podSet := fromObject(&podInGroup).PodSets()
+			podSet[0].Name = roleHash
+
+			resultPodSets = append(resultPodSets, podSet[0])
+		}
+	}
+
+	return resultPodSets, nil
+}
+
+func (p *Pod) ConstructComposableWorkload(ctx context.Context, c client.Client, r record.EventRecorder) (*kueue.Workload, error) {
+	log := ctrl.LoggerFrom(ctx)
+	object := p.Object()
+
+	var (
+		podSets     []kueue.PodSet
+		podsInGroup corev1.PodList
+	)
+
+	if p.groupName() == "" {
+		podSets = p.PodSets()
+	} else {
+		if err := c.List(ctx, &podsInGroup, client.MatchingLabels{
+			GroupNameLabel: p.groupName(),
+		}); err != nil {
+			return nil, err
+		}
+
+		groupTotalCount, err := p.groupTotalCount()
+		if err != nil {
+			return nil, err
+		}
+
+		if len(podsInGroup.Items) != groupTotalCount {
+			r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", errMsgIncorrectTotalGroupCount)
+			return nil, fmt.Errorf(errMsgIncorrectTotalGroupCount)
```

Comment:
if you return an error, the Pod sync will be retried immediately. We don't want that, we would only want to retry once a new pod is observed.

To do this, we might need to create an `UnretriableError` struct that we can ignore from the generic reconciler. Somewhat similar to how we ignore `NotFound` errors.

Reply:
JFYI we could use the [`TerminalError`](https://github.com/kubernetes-sigs/controller-runtime/blob/06fb2e9592987c3182bae373f83e4b7e6f0b9e2b/pkg/reconcile/reconcile.go#L113) from `controller-runtime` here, but I will implement a custom error, since we don't need those types of errors coming up in metrics.

Reply:
Added UnretryableError in https://github.com/kubernetes-sigs/kueue/pull/1319/commits/fb5a4eb591521e5e3e2d7acfb4b6268a697b52e9